fdbserver/TagPartitionedLogSystem.actor.cpp

```diff
@@ -2041,40 +2047,128 @@ ACTOR Future<Void> TagPartitionedLogSystem::getDurableVersionChanged(LogLockInfo
 	return Void();
 }
 
+void getTLogLocIds(std::vector<Reference<LogSet>>& tLogs,
+                   std::vector<std::tuple<int, std::vector<TLogLockResult>>>& logGroupResults,
+                   std::vector<std::vector<int>>& tLogLocIds,
+                   int& maxTLogLocId) {
+	// Initialization.
+	tLogLocIds.clear();
+	tLogLocIds.resize(logGroupResults.size());
+	maxTLogLocId = std::numeric_limits<int>::min();
+
+	// Map the interfaces of all (local) tLogs to their corresponding locations in LogSets.
+	std::map<UID, int> interfLocMap;
+	int location = 0;
+	for (auto& it : tLogs) {
+		if (!it->isLocal) {
+			continue;
+		}
+		for (int i = 0; i < it->logServers.size(); i++) {
+			interfLocMap[it->logServers[i]->get().interf().id()] = location++;
+		}
+	}
+
+	// Find the locations of tLogs in "logGroupResults".
+	int logGroupId = 0;
+	for (auto& logGroupResult : logGroupResults) {
+		for (auto& tLogResult : std::get<1>(logGroupResult)) {
+			ASSERT(interfLocMap.find(tLogResult.logId) != interfLocMap.end());
+			tLogLocIds[logGroupId].push_back(interfLocMap[tLogResult.logId]);
+			maxTLogLocId = std::max(maxTLogLocId, interfLocMap[tLogResult.logId]);
+		}
+		logGroupId++;
+	}
+}
+
+void populateBitset(boost::dynamic_bitset<>& bs, std::vector<int>& ids) {
+	for (auto& id : ids) {
+		bs.set(id);
+	}
+}
+
 // If VERSION_VECTOR_UNICAST is enabled, one tLog's DV may advance beyond the min(DV) over all tLogs.
 // This function finds the highest recoverable version for each tLog group over all log groups.
 // All prior versions to the chosen RV must also be recoverable.
 // TODO: unit tests to stress UNICAST
-Version getRecoverVersionUnicast(std::vector<std::tuple<int, std::vector<TLogLockResult>>>& logGroupResults,
+Version getRecoverVersionUnicast(std::vector<Reference<LogSet>>& logServers,
```

Comment:
Regarding "minEnd": Shouldn't we be using "knownCommittedVersion" as "minEnd" (that is, minEnd = std::get<0>getDurableVersion())? @dlambrig maybe you can comment on this, thanks!

Reply:
`minEnd` is the RV calculated using the non-version-vector algorithm. Originally, the idea had been to treat that version as a floor. So function `getRecoverVersionUnicast ` could increase the RV, but not choose a value below it. We should reconsider this.

Reply:
> So function getRecoverVersionUnicast could increase the RV, but not choose a value below it.

I think it can choose a value below it because all versions till "minEnd" are not guaranteed have committed (when unicast is enabled).

Reply:
We discussed this part, it is probably fine to use "min(DV)" here.

Reply:
Modified code to use "min(KCV)" as the default recovery version and added an assert that verifies that the computed recovery version is greater than or equal to "min(DV)".